validator/src/journal/block_scheduler.rs

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright 2018 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ------------------------------------------------------------------------------
+ */
+
+use std::collections::{HashMap, HashSet};
+use std::sync::RwLock;
+
+use block::Block;
+use journal::block_manager::BlockManager;
+use journal::block_wrapper::BlockStatus;
+use metrics;
+
+lazy_static! {
+    static ref COLLECTOR: metrics::MetricsCollectorHandle =
+        metrics::get_collector("sawtooth_validator.block_validator");
+}
+
+pub struct BlockScheduler<F: Fn(&str) -> BlockStatus> {
+    state: RwLock<BlockSchedulerState<F>>,
+}
+
+impl<F: Fn(&str) -> BlockStatus> BlockScheduler<F> {
+    pub fn new(block_manager: BlockManager) -> Self {
+        BlockScheduler {
+            state: RwLock::new(BlockSchedulerState {
+                block_manager,
+                block_validity_fn: None,
+                pending: HashSet::new(),
+                processing: HashSet::new(),
+                descendants: HashMap::new(),
+            }),
+        }
+    }
+
+    pub fn set_block_validity_fn(&self, func: F) {
+        self.state
+            .write()
+            .expect("The BlockScheduler RwLock was poisoned")
+            .set_block_validity_fn(func);
+    }
+
+    pub fn schedule(&self, blocks: Vec<Block>) -> Vec<Block> {
+        self.state
+            .write()
+            .expect("The BlockScheduler RwLock was poisoned")
+            .schedule(blocks)
+    }
+
+    /// Mark the block as having completed Block Validation and return any blocks
+    /// that are now available.
+    pub fn done(&self, block: Block) -> Vec<Block> {
+        self.state
+            .write()
+            .expect("The BlockScheduler RwLock was poisoned")
+            .done(block)
+    }
+
+    pub fn contains(&self, block_id: &str) -> bool {
+        self.state
+            .read()
+            .expect("The BlockScheduler RwLock was poisoned")
+            .contains(block_id)
+    }
+}
+
+struct BlockSchedulerState<F: Fn(&str) -> BlockStatus> {
+    pub block_manager: BlockManager,
+    pub block_validity_fn: Option<F>,
+    pub pending: HashSet<String>,
+    pub processing: HashSet<String>,
+    pub descendants: HashMap<String, Vec<Block>>,
+}
+
+impl<F: Fn(&str) -> BlockStatus> BlockSchedulerState<F> {
+    fn set_block_validity_fn(&mut self, func: F) {
```

Comment:
Secondly, who is providing this method? Are you going to run in to a deadlock with other locks?

Reply:
That method comes from the ChainController. The same as is used now in the BlockScheduler.